### PR TITLE
feat: Add new lockfile format

### PR DESCRIPTION
**Pull request by bartlomieju**
Introduces a new lockfile format that will be used to support locking
"npm" dependencies.

Currently the format looks as follows:

```
// This file is automatically generated by Deno, do not edit its contents
// manually. This file should be commited to your repository.
{
  "version": "2",
  "remote": {
    "https://deno.land/std@0.160.0/http/server.ts": "asdwetsw44523asdfgfas..",
    "https://deno.land/std@0.160.0/http/file_server.ts": "asdwetsw44523asdfgfas.."
  }
}

```

A follow up PR will add "npm" key that will be used to store information related
to "npm" dependencies and their resolution.

The new format is used when `--lock-write` is present, if user tries to load
a lock file using the old format it will still work.

### CODE REVIEW

File: cli/lockfile.rs [hunk 1]
```diff
@@ -30,19 +45,41 @@ pub struct Lockfile {
...
+        version: "2".to_string(),
+        remote: BTreeMap::new(),
+      })
     } else {
       let s = std::fs::read_to_string(&filename).with_context(|| {
         format!("Unable to read lockfile: {}", filename.display())
       })?;
-      serde_json::from_str(&s)
-        .context("Unable to parse contents of the lockfile")?
+      let value: serde_json::Value = jsonc_parser::parse_to_serde_value(
```
Comment (dsherret) [inline cli/lockfile.rs:58]: We should probably keeps this using serde_json. Allowing jsonc for the lockfile doesn't seem necessary?

Comment (dsherret) [reply cli/lockfile.rs:58]: Oh, I see it's for the leading comments. The one issue with having comments is then it makes it harder for some other tools to read the lockfile. I think what's said in the comments is somewhat self explanatory with a lockfile?

**Review by dsherret — APPROVED** [review]
LGTM, but jsonc is questionable.

File: cli/lockfile.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/lockfile.rs:58]: Yeah, maybe, but other tools usually add such comments (though they often use a file format that supports comments OOTB

**Comment by bartlomieju** [comment]
Reverted to using JSON

File: cli/lockfile.rs [hunk 2]
```diff
@@ -30,19 +45,41 @@ pub struct Lockfile {
...
+      )
+      .context("Unable to parse contents of the lockfile")?
+      .unwrap_or_default();
+      let version = value.get("version").and_then(|v| v.as_str());
+      if version == Some("2") {
+        let content: LockfileV2Content =
+          serde_json::from_value(value).context("Unable to parse lockfile")?;
+        LockfileContent::V2(content)
+      } else {
+        let content: LockfileV1Content =
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/lockfile.rs:66]: Let's discuss it in a follow up

**Pull request merged**